jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java

```diff
@@ -571,61 +574,67 @@ static Chunk next(Chunk chunk)
-         * source is {@link #isTerminal() terminal} in which case {@code this} is returned, or
-         * if {@code position == limit} in which case {@link #EOF} or {@link #EMPTY} is
+         * source {@link #hasRemaining() has no remaining bytes} in which case:</p>
+         * <ul>
+         * <li>{@code this} is returned if {@link #isLast()} is {@code true}</li>
+         * <li>{@link #EMPTY} is returned if {@link #isLast()} is {@code false}</li>
+         * </ul>
+         * <p>If {@code position == limit} then either {@link #EOF} or {@link #EMPTY} is
          * returned depending on the value of {@code last}.</p>
-         * <p>The returned {@code Chunk} retains the source {@code Chunk} and it is linked
-         * to it via {@link #from(ByteBuffer, boolean, Retainable)}.</p>
+         * <p>If the source {@code Chunk} has remaining bytes, the returned {@code Chunk}
+         * is linked to the source {@code Chunk} via
+         * {@link #from(ByteBuffer, boolean, Retainable)}.</p>
          *
          * @param position the position at which the slice begins
          * @param limit the limit at which the slice ends
          * @param last whether the new Chunk is last
-         * @return a new {@code Chunk} retained from the source {@code Chunk} with a slice
-         * of the source {@code Chunk}'s {@code ByteBuffer}
+         * @return a new {@code Chunk} with a slice of the source {@code Chunk}'s
+         * {@code ByteBuffer}
          */
         default Chunk slice(int position, int limit, boolean last)
         {
-            if (isTerminal())
-                return this;
-            if (position == limit)
-                return last ? EOF : EMPTY;
-            ByteBuffer sourceBuffer = getByteBuffer();
-            int sourceLimit = sourceBuffer.limit();
-            sourceBuffer.limit(limit);
-            int sourcePosition = sourceBuffer.position();
-            sourceBuffer.position(position);
-            ByteBuffer slice = sourceBuffer.slice();
-            sourceBuffer.limit(sourceLimit);
-            sourceBuffer.position(sourcePosition);
-            retain();
-            return from(slice, last, this);
+            if (hasRemaining())
+            {
+                if (position == limit)
+                    return last ? EOF : EMPTY;
+                ByteBuffer sourceBuffer = getByteBuffer();
+                int sourceLimit = sourceBuffer.limit();
+                sourceBuffer.limit(limit);
+                int sourcePosition = sourceBuffer.position();
+                sourceBuffer.position(position);
+                ByteBuffer slice = sourceBuffer.slice();
+                sourceBuffer.limit(sourceLimit);
+                sourceBuffer.position(sourcePosition);
+                return from(slice, last, this);
+            }
+            else
+            {
+                return isLast() ? this : EMPTY;
+            }
```

Comment:
This method could just be (from #9141 ):
```suggestion
            {
                ByteBuffer sourceBuffer = getByteBuffer();
                if (position == limit)
                    return last ? EOF : EMPTY;
                return from(sourceBuffer.slice(position, limit - position), last, this);
            }
```